### ISSUE TITLE
[Inductor][ROCm] Comprehensive padding is skipped when an intermediate buffer is frozen under a user-visible consumer

### ISSUE THREAD

**Issue by yoshua2077** [pytorch/pytorch#197870]
### Bug description

On ROCm/HIP, TorchInductor can incorrectly skip comprehensive padding for an intermediate tensor.

The intermediate buffer is produced by `add`, but it is frozen while the current FX node is its consumer `mm`. Since the final `mm` is a user-visible output and has `dislike_padding=True`, `_pad_strides()` incorrectly applies the consumer's padding preference to the producer buffer.

### Minimal reproduction

```python
import torch
from torch._inductor import config, metrics


def f(x, y):
    x = torch.matmul(x, y)
    x = x + 1
    return torch.matmul(x, y)


torch.manual_seed(0)

x = torch.randn(2, 16, 2047, device="cuda")
y = torch.randn(2047, 2047, device="cuda")

ref = f(x, y)

metrics.reset()

with config.patch(
    {
        "comprehensive_padding": True,
        "padding_alignment_bytes": 128,
        "padding_stride_threshold": 1024,
        "triton.cudagraphs": False,
    }
):
    compiled = torch.compile(f, fullgraph=True)
    out = compiled(x, y)

torch.testing.assert_close(ref, out, atol=1e-2, rtol=1e-2)

print("num_comprehensive_padding:", metrics.num_comprehensive_padding)
assert metrics.num_comprehensive_padding > 0
```

### Actual behavior

```text
num_comprehensive_padding: 0
AssertionError
```

The intermediate `[2, 16, 2047]` tensor has stride:

```text
[32752, 2047, 1]
```

With float32 and 128-byte alignment, it should be padded to:

```text
[32768, 2048, 1]
```

but the original stride is kept.

### Expected behavior

The intermediate tensor should be comprehensively padded, while the final user-visible output should still preserve its required stride.

`metrics.num_comprehensive_padding` should therefore be greater than zero.

### Root cause

Tracing the layout freeze shows:

```text
buffer: buf1
producer: add
producer target: aten.add.Tensor
producer dislike_padding: False

current node: mm_1
current target: aten.mm.default
current dislike_padding: True
```

When `_pad_strides()` runs:

```text
current node: mm_1
current dislike_padding: True
input stride:  [32752, 2047, 1]
output stride: [32752, 2047, 1]
```

So the buffer belongs to `add`, but the padding decision is made using the consumer `mm_1`.

`Layout._pad_strides()` currently uses `V.get_current_node()` to check `dislike_padding`, which in this case refers to the consumer rather than the origin node of the buffer being frozen.

### A/B verification

As a local experiment, running the layout freeze under the buffer's origin FX node:

```python
origin = buf.get_origin_node()

with V.set_current_node(origin):
    ...
```

changes the result to:

```text
num_comprehensive_padding: 1
PASS
```

No other behavior was changed.

### Proposed fix location

The relevant location appears to be:

```text
torch/_inductor/ir.py
ExternKernel.convert_to_reinterpret_view()
```

This function already obtains:

```python
x_unwrap_view_fx_node = buf.get_origin_node()
```

but layout freezing still runs under the existing `V.current_node`.

Using the buffer's origin node as the current node while calling `freeze_layout()` / `freeze_layout_with_same_order()` appears to avoid incorrectly inheriting `dislike_padding` from the consumer.

### Environment

- PyTorch: 2.14.0
- Backend: ROCm / HIP
- `torch.compile` / TorchInductor
- AMD GPU

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben

**Comment by Abdellox** [comment]
Thanks for reporting this! I'd like to help investigate.

Could you provide:
- Steps to reproduce
- Expected vs actual behavior
- Environment details

I'll look into the codebase and see if I can identify the root cause.

**Comment by yoshua2077** [comment]
> Thanks for reporting this! I'd like to help investigate.
> 
> Could you provide:
> 
> * Steps to reproduce
> * Expected vs actual behavior
> * Environment details
> 
> I'll look into the codebase and see if I can identify the root cause.

Thanks for taking a look. The issue body already includes a standalone minimal reproduction, expected vs. actual behavior, environment details, the relevant trace, and an A/B result showing that switching the layout-freeze context to the buffer's origin FX node changes num_comprehensive_padding from 0 to 1. If useful, I can also provide a small regression-test patch against test/inductor/test_padding.py.

**Comment by yoshua2077** [comment]
I validated a minimal source change locally.

The fix is to run the layout-freeze block in `ExternKernel.convert_to_reinterpret_view()` under the origin FX node of the buffer being frozen, rather than the consumer that happens to be the current lowering node:

```diff
diff --git a/torch/_inductor/ir.py b/torch/_inductor/ir.py
@@
         x_unwrap_view_fx_node = buf.get_origin_node()
-        # Prefer channels last format according to how the format is set from eager.
-        if (
-            x_unwrap_view_fx_node is not None
-            ...
-        ):
-            x_unwrap_view.freeze_layout_with_same_order(...)
-        else:
-            x_unwrap_view.freeze_layout()
+        padding_fx_node = x_unwrap_view_fx_node or V.get_current_node()
+        with V.set_current_node(padding_fx_node):
+            # Prefer channels last format according to how the format is set from eager.
+            if (
+                x_unwrap_view_fx_node is not None
+                ...
+            ):
+                x_unwrap_view.freeze_layout_with_same_order(...)
+            else:
+                x_unwrap_view.freeze_layout()
```

With only that context change:

```text
standalone repro:
num_comprehensive_padding: 0 -> 1

test/inductor/test_padding.py::PaddingTest::test_pad_3d_tensor:
PASSED

test/inductor/test_padding.py::PaddingTest:
47 passed, 4 skipped
```

The producer/consumer trace also changes as expected: before the fix, the buffer producer is `add` but `_pad_strides()` sees current node `mm_1` with `dislike_padding=True`; under the producer context, the intermediate stride is padded from `[32752, 2047, 1]` to `[32768, 2048, 1]`.

This suggests the fix can stay localized to `torch/_inductor/ir.py::ExternKernel.convert_to_reinterpret_view()` without changing the padding policy itself.

**Comment by yoshua2077** [comment]
One more data point: I reproduced the same producer/consumer-context behavior on both NVIDIA and AMD/DCU backends.

Baseline on both backends:

```text
current node = mm_1
mm_1.dislike_padding = True
[32752, 2047, 1] -> unchanged
FAIL
```

When the freeze is evaluated under the actual producer context:

```text
origin = add
add.dislike_padding = False
[32752, 2047, 1] -> [32768, 2048, 1]
PASS
```

So this does not appear to be AMD/HIP/Triton-specific. The same Inductor semantics are involved on both backends.

More precisely, `ExternKernel.convert_to_reinterpret_view()` is freezing the layout of an intermediate buffer produced by `add`, but `_pad_strides()` observes the current consumer `mm_1` through `V.current_node`. Since the user-visible `mm_1` is marked `dislike_padding=True`, that property is incorrectly applied to the earlier `add` buffer and suppresses comprehensive padding.

Switching `V.current_node` to the buffer's origin node only for the layout-freeze block restores the expected padding behavior on both backends.
